cmd/disk-cache-backend.go

```diff
@@ -507,12 +587,62 @@ func (c *diskCache) bitrotReadFromCache(ctx context.Context, filePath string, of
 	return nil
 }
 
+// decryptCacheObjectETag tries to decrypt the ETag saved in encrypted format using the cache KMS
+func decryptCacheObjectETag(info *ObjectInfo) (encrypted bool, objectEncryptionKey []byte, err error) {
+	// Directories are never encrypted.
+	if info.IsDir {
+		return false, nil, nil
+	}
+	encrypted = crypto.S3.IsEncrypted(info.UserDefined) && isCacheEncrypted(info.UserDefined)
+
+	switch {
+	default:
+		return false, nil, nil
+	case encrypted:
+		if globalCacheKMS == nil {
+			return true, nil, errKMSNotConfigured
+		}
+		keyID, kmsKey, sealedKey, err := crypto.S3.ParseMetadata(info.UserDefined)
+
+		if err != nil {
+			return true, nil, err
+		}
+		extKey, err := globalCacheKMS.UnsealKey(keyID, kmsKey, crypto.Context{info.Bucket: path.Join(info.Bucket, info.Name)})
+		if err != nil {
+			return true, nil, err
+		}
+		var objectKey crypto.ObjectKey
+		if err = objectKey.Unseal(extKey, sealedKey, crypto.S3.String(), info.Bucket, info.Name); err != nil {
+			return true, nil, err
+		}
+		info.ETag = tryDecryptCacheETag(objectKey[:], info.ETag, false)
+	}
+
+	return
+}
+
+// helper to decrypt Etag given object encryption key and encrypted ETag
+func tryDecryptCacheETag(key []byte, encryptedETag string, ssec bool) string {
+
+	var objectKey crypto.ObjectKey
+	copy(objectKey[:], key)
+	encBytes, err := hex.DecodeString(encryptedETag)
+	if err != nil {
+		return encryptedETag
+	}
+	etagBytes, err := objectKey.UnsealETag(encBytes)
+	if err != nil {
+		return encryptedETag
+	}
+	return string(etagBytes)
+}
+
 // Get returns ObjectInfo and reader for object from disk cache
 func (c *diskCache) Get(ctx context.Context, bucket, object string, rs *HTTPRangeSpec, h http.Header, opts ObjectOptions) (gr *GetObjectReader, err error) {
```

Comment:
Should we not rely on `SSECacheEncrypted` for decrypting the cache data? AFAICU, We are using a seperate master key for cache - `MINIO_CACHE_ENCRYPTION_MASTER_KEY`. So, there can be situations where SSE is enabled but cache encryption is not. right ?..  So i guess we cannot rely on `crypto.IsEncrypted()` (in `NewGetObjectReader`) alone to decide whether the underlying cache data is encrypted or not. And there are possibilities where we could mistakenly try to decrypt cache data with server side KMS creds.

So i guess the only way to trustfully check if the cache data is encrypted or not is through `SSECacheEncrypted`. Correct me if i'm missing anything.

Reply:
Ok.. Looks like we are not preserving the sensitive metadata entries in cache metadata.. Ignore the comment